### PR TITLE
Make time of day methods handle daylight savings time (DST) better

### LINKED ISSUE #5244: `minuteOfDay` returns surprising results on DST days
A client reported a daylight savings time (DST) problem with `minuteOfDay`.

He expected the following to be minute=1,020 but it gives minute=960 due to the clock change (we lost an hour). 

```python
empty_table(1).update("time='2024-03-10T17:00:00.000 ET'").update("minute=minuteOfDay(time, timeZone(`ET`))")
```

### PR DESCRIPTION
**Pull request by chipkent**
Change time of day methods to handle daylight savings time better.

The old method signatures are deprecated and need to be removed in a version or two.

Resolves #5244

### CODE REVIEW

File: engine/time/src/main/java/io/deephaven/time/DateTimeUtils.java [hunk 1]
```diff
@@ -2567,14 +2569,214 @@ public static long nanosOfDay(@Nullable final Instant instant, @Nullable final Z
...
+     * @param instant time
+     * @param timeZone time zone
+     * @return {@link QueryConstants#NULL_INT} if either input is {@code null}; otherwise, number of milliseconds that
+     *         have elapsed since the top of the day
+     * @deprecated Use {@link #millisOfDay(Instant, ZoneId, boolean)} instead.  To be removed soon.
+     */
+    @ScriptApi
+    @Deprecated
+    public static int millisOfDay(@Nullable final Instant instant, @Nullable final ZoneId timeZone) {
+        return millisOfDay(instant, timeZone, false);
```
Comment (kosak) [inline engine/time/src/main/java/io/deephaven/time/DateTimeUtils.java:2596]: I don't know what the culture is here, but I'm feeling a little sad that we have to provide all these `${UNIT}OfDay` methods. Would it be ok just to provide `nanosOfDay` and let our user call `nanosTo${UNIT}` e.g. `nanosToSeconds` or whatever?

Just an observation, I'm not deeply opinionated on this.

File: engine/time/src/main/java/io/deephaven/time/DateTimeUtils.java [hunk 2]
```diff
@@ -2567,14 +2569,214 @@ public static long nanosOfDay(@Nullable final Instant instant, @Nullable final Z
...
+     * @deprecated Use {@link #hourOfDay(ZonedDateTime, boolean)} instead.  To be removed soon.
+     */
+    @ScriptApi
+    @Deprecated
+    public static int hourOfDay(@Nullable final ZonedDateTime dateTime) {
+        return hourOfDay(dateTime, false);
+    }
+
+    ///////////////////////////////////////////////////////////////////////////////////////////////////
+    // ^^^^^^^ DEPRECATED METHODS ^^^^^^^
```
Comment (kosak) [inline engine/time/src/main/java/io/deephaven/time/DateTimeUtils.java:2729]: Consider:
```
↑↑↑↑↑↑↑ THE METHODS ABOVE ARE DEPRECATED AND WILL BE REMOVED SOON ↑↑↑↑↑↑↑
```

```
⇑⇑⇑⇑⇑⇑⇑⇑⇑
⇡⇡⇡⇡⇡⇡⇡⇡
↥↥↥↥↥↥↥↥↥
```

File: engine/time/src/main/java/io/deephaven/time/DateTimeUtils.java [hunk 3]
```diff
@@ -2594,166 +2796,246 @@ public static long nanosOfDay(@Nullable final LocalTime localTime) {
...
      * Returns the number of milliseconds that have elapsed since the top of the day.
-     * <p>
-     * On days when daylight savings time events occur, results may be different from what is expected based upon the
-     * local time. For example, on daylight savings time change days, 9:30AM may be earlier or later in the day based
-     * upon if the daylight savings time adjustment is forwards or backwards.
      *
      * @param instant time
      * @param timeZone time zone
+     * @param localTime if {@code true}, returns the number of nanos from the start of the day according to the local
+     *        time. In this case, 9:30AM always returns the same value. If {@code false}, returns the number of nanos
```
Comment (kosak) [inline engine/time/src/main/java/io/deephaven/time/DateTimeUtils.java:2803]: nanos -> millis

File: engine/time/src/main/java/io/deephaven/time/DateTimeUtils.java [hunk 4]
```diff
@@ -2594,166 +2796,246 @@ public static long nanosOfDay(@Nullable final LocalTime localTime) {
...
     /**
      * Returns the number of milliseconds that have elapsed since the top of the day.
-     * <p>
-     * On days when daylight savings time events occur, results may be different from what is expected based upon the
-     * local time. For example, on daylight savings time change days, 9:30AM may be earlier or later in the day based
-     * upon if the daylight savings time adjustment is forwards or backwards.
      *
      * @param dateTime time
+     * @param localTime if {@code true}, returns the number of nanos from the start of the day according to the local
+     *        time. In this case, 9:30AM always returns the same value. If {@code false}, returns the number of nanos
```
Comment (kosak) [inline engine/time/src/main/java/io/deephaven/time/DateTimeUtils.java:2826]: nanos -> millis

File: engine/time/src/main/java/io/deephaven/time/DateTimeUtils.java [hunk 5]
```diff
@@ -2594,166 +2796,246 @@ public static long nanosOfDay(@Nullable final LocalTime localTime) {
...
     /**
      * Returns the number of seconds that have elapsed since the top of the day.
-     * <p>
-     * On days when daylight savings time events occur, results may be different from what is expected based upon the
-     * local time. For example, on daylight savings time change days, 9:30AM may be earlier or later in the day based
-     * upon if the daylight savings time adjustment is forwards or backwards.
      *
      * @param dateTime time
+     * @param localTime if {@code true}, returns the number of nanos from the start of the day according to the local
+     *        time. In this case, 9:30AM always returns the same value. If {@code false}, returns the number of nanos
```
Comment (kosak) [inline engine/time/src/main/java/io/deephaven/time/DateTimeUtils.java:2887]: nanos->seconds

File: engine/time/src/main/java/io/deephaven/time/DateTimeUtils.java [hunk 6]
```diff
@@ -2594,166 +2796,246 @@ public static long nanosOfDay(@Nullable final LocalTime localTime) {
...
      * Returns the number of minutes that have elapsed since the top of the day.
-     * <p>
-     * On days when daylight savings time events occur, results may be different from what is expected based upon the
-     * local time. For example, on daylight savings time change days, 9:30AM may be earlier or later in the day based
-     * upon if the daylight savings time adjustment is forwards or backwards.
      *
      * @param instant time
      * @param timeZone time zone
+     * @param localTime if {@code true}, returns the number of nanos from the start of the day according to the local
+     *        time. In this case, 9:30AM always returns the same value. If {@code false}, returns the number of nanos
```
Comment (kosak) [inline engine/time/src/main/java/io/deephaven/time/DateTimeUtils.java:2925]: nanos->minutes

File: engine/time/src/main/java/io/deephaven/time/DateTimeUtils.java [hunk 7]
```diff
@@ -2594,166 +2796,246 @@ public static long nanosOfDay(@Nullable final LocalTime localTime) {
...
     /**
      * Returns the number of minutes that have elapsed since the top of the day.
-     * <p>
-     * On days when daylight savings time events occur, results may be different from what is expected based upon the
-     * local time. For example, on daylight savings time change days, 9:30AM may be earlier or later in the day based
-     * upon if the daylight savings time adjustment is forwards or backwards.
      *
      * @param dateTime time
+     * @param localTime if {@code true}, returns the number of nanos from the start of the day according to the local
+     *        time. In this case, 9:30AM always returns the same value. If {@code false}, returns the number of nanos
```
Comment (kosak) [inline engine/time/src/main/java/io/deephaven/time/DateTimeUtils.java:2948]: nanos->minutes

File: engine/time/src/main/java/io/deephaven/time/DateTimeUtils.java [hunk 8]
```diff
@@ -2594,166 +2796,246 @@ public static long nanosOfDay(@Nullable final LocalTime localTime) {
...
      * Returns the number of hours that have elapsed since the top of the day.
-     * <p>
-     * On days when daylight savings time events occur, results may be different from what is expected based upon the
-     * local time. For example, on daylight savings time change days, 9:30AM may be earlier or later in the day based
-     * upon if the daylight savings time adjustment is forwards or backwards.
      *
      * @param instant time
      * @param timeZone time zone
+     * @param localTime if {@code true}, returns the number of nanos from the start of the day according to the local
+     *        time. In this case, 9:30AM always returns the same value. If {@code false}, returns the number of nanos
```
Comment (kosak) [inline engine/time/src/main/java/io/deephaven/time/DateTimeUtils.java:2986]: nanos->hours

File: engine/time/src/main/java/io/deephaven/time/DateTimeUtils.java [hunk 9]
```diff
@@ -2594,166 +2796,246 @@ public static long nanosOfDay(@Nullable final LocalTime localTime) {
...
     /**
      * Returns the number of hours that have elapsed since the top of the day.
-     * <p>
-     * On days when daylight savings time events occur, results may be different from what is expected based upon the
-     * local time. For example, on daylight savings time change days, 9:30AM may be earlier or later in the day based
-     * upon if the daylight savings time adjustment is forwards or backwards.
      *
      * @param dateTime time
+     * @param localTime if {@code true}, returns the number of nanos from the start of the day according to the local
+     *        time. In this case, 9:30AM always returns the same value. If {@code false}, returns the number of nanos
```
Comment (kosak) [inline engine/time/src/main/java/io/deephaven/time/DateTimeUtils.java:3009]: nanos->hours

**Review by kosak — COMMENTED** [review]
just some small comments

**Review by kosak — APPROVED** [review]

**Comment by deephaven-internal** [comment]
Labels indicate documentation is required. Issues for documentation have been opened:

Community: https://github.com/deephaven/deephaven-docs-community/issues/172

**Pull request merged**